src/libexpr/flake/call-flake.nix

```diff
@@ -12,7 +12,48 @@ overrides:
 
 let
 
-  lockFile = builtins.fromJSON lockFileStr;
+  rawLockFile = builtins.fromJSON lockFileStr;
+
+  # Given a function that returns a string key that sufficiently
+  # identifies an attribute value, return an attribute set that
+  # returns a unique attribute name for all equivalent attribute values.
+  deduplicationMap = f: m:
+    builtins.listToAttrs (builtins.attrValues (builtins.mapAttrs
+      (name: value: { name = f value; value = name; })
+      m
+    ));
+
+  # A node is considered duplicate if its `locked` value is
+  # equivalent, unless `nodeKey` was overridden.
+  nodeDeduplicationKey = { nodeKey, node }:
+    if overrides ? ${nodeKey}
+    then builtins.toJSON { overriddenAs = nodeKey; }
+    else builtins.toJSON node.locked;
```

Comment:
I think that's safe to assume, and I'm not too worried because it wouldn't fail silently.

However, I think you do need to use the `inputs`, because the same flake source will behave differently depending on its inputs.
Pulling that off is bit tricky because a naive implementation leads to an infinite recursion. You'd have to do it iteratively, starting at the leaves and then work your way to the root. I don't know if such complexity is worth the risk right now, and it might not even catch that many extra duplicates.

Reply:
Are you saying the current approach is incorrect because the deduplication key doesn't take into account `inputs`?

Uh, this is problematic indeed. I can't recall how I came to the conclusion that this was fine, especially since it was something I handled in the initial commit before rewriting it. I think I was trying to avoid eagerly evaluating inputs, but I might have had issues with infinite recursion as well.

I'll have another look if this can be fixed in a way that isn't too complex. Otherwise it's probably better to do this using some sort of memoization primop (#10280).

Reply:
Actually, I'm not sure it's possible to avoid both infinite recursion and eager evaluation.